File: Web/CustomPreview/C#/CustomPreview/RdlxHandler.cs
using System;
using System.IO;
using System.Web;
using System.Web.Caching;
using GrapeCity.ActiveReports.Export.Html.Page;
using GrapeCity.ActiveReports.Extensibility.Rendering;
using GrapeCity.ActiveReports.Rendering.IO;

namespace GrapeCity.ActiveReports.Samples.Web.CustomPreview
{
	public class RdlxHandler : IHttpHandler
	{
		private const string HandlerExtension = ".rdlx";
		private const string HandlerCacheExtension = ".rdlxWeb";

		public void ProcessRequest(HttpContext context)
		{
			if (!context.Request.Url.AbsolutePath.EndsWith(HandlerExtension))
			{
				if (!context.Request.Url.AbsolutePath.EndsWith(HandlerCacheExtension))
					return;

				// return image
				var keyName = Path.GetFileName(context.Request.FilePath);
				var cacheItem = context.Cache[keyName];
				context.Response.BinaryWrite((byte[]) cacheItem);
				return;
			}

			context.Response.ContentType = "text/html";
			var streams = new HtmlStreamProvider();
			try
			{
				using (var report = new PageReport(new FileInfo(context.Server.MapPath(context.Request.Url.LocalPath))))
				{
					var html = new HtmlRenderingExtension();
					var settings = (Settings)html.GetSupportedSettings();
					settings.StyleStream = false;
					settings.OutputTOC = false;
					settings.Mode = RenderMode.Paginated;
					report.Document.Render(html, streams, ((ISettings)settings).GetSettings());
				}

				foreach (var secondaryStreamInfo in streams.GetSecondaryStreams())
				{
					var secondaryStream = (MemoryStream) secondaryStreamInfo.OpenStream();
					// 30 seconds should be enough to load all images
					context.Cache.Insert(secondaryStreamInfo.Uri.OriginalString, secondaryStream.ToArray(), null,
						Cache.NoAbsoluteExpiration, new TimeSpan(0, 0, 30));
				}
			}
			catch (ReportException eRunReport)
			{
				// Failure running report, just report the error to the user.
				context.Response.Write(Properties.Resource.Error);
				context.Response.Write(eRunReport.ToString());
				return;
			}

			var primaryStream = (MemoryStream) streams.GetPrimaryStream().OpenStream();
			context.Response.BinaryWrite(primaryStream.ToArray());
		}

		public bool IsReusable
		{
			get { return true; }
		}

		private sealed class HtmlStreamProvider : MemoryStreamProvider
		{
			protected override Uri GetStreamUri(string extension)
			{
				Uri uri = base.GetStreamUri(extension);
				return new Uri(uri.OriginalString + HandlerCacheExtension, UriKind.Relative);
			}
		}
	}
}

File: Web/CustomPreview/VB.NET/CustomPreview/RdlxHandler.vb
Imports System.IO
Imports System.Web
Imports System.Web.Caching
Imports GrapeCity.ActiveReports.Export.Html.Page
Imports GrapeCity.ActiveReports.Extensibility.Rendering
Imports GrapeCity.ActiveReports.Extensibility.Rendering.IO
Imports GrapeCity.ActiveReports.Rendering.IO

Public Class RdlxHandler
	Implements IHttpHandler
	Private Const HandlerExtension As String = ".rdlx"
	Private Const HandlerCacheExtension As String = ".rdlxWeb"

	Public Sub ProcessRequest(context As HttpContext) Implements IHttpHandler.ProcessRequest
		If Not context.Request.Url.AbsolutePath.EndsWith(HandlerExtension) Then
			If Not context.Request.Url.AbsolutePath.EndsWith(HandlerCacheExtension) Then
				Return
			End If

			' return image
			Dim keyName As String = Path.GetFileName(context.Request.FilePath)
			Dim cacheItem As Byte() = DirectCast(context.Cache(keyName), Byte())
			context.Response.BinaryWrite(cacheItem)
			Return
		End If

		context.Response.ContentType = "text/html"
		Dim streams As HtmlStreamProvider = New HtmlStreamProvider()
		Try
			Using report As PageReport = New PageReport(New FileInfo(context.Server.MapPath(context.Request.Url.LocalPath)))
				Dim html As HtmlRenderingExtension = New HtmlRenderingExtension()
				Dim settings As Settings = DirectCast(html.GetSupportedSettings(), Settings)
				settings.StyleStream = False
				settings.OutputTOC = False
				settings.Mode = RenderMode.Paginated
				report.Document.Render(html, streams, DirectCast(settings, ISettings).GetSettings())
			End Using

			For Each secondaryStreamInfo As StreamInfo In streams.GetSecondaryStreams()
				Dim secondaryStream As MemoryStream = DirectCast(secondaryStreamInfo.OpenStream(), MemoryStream)
				' 30 seconds should be enough to load all images
				context.Cache.Insert(secondaryStreamInfo.Uri.OriginalString, secondaryStream.ToArray(), Nothing, Cache.NoAbsoluteExpiration, New TimeSpan(0, 0, 30))
			Next
		Catch eRunReport As ReportException
			' Failure running report, just report the error to the user.
			context.Response.Write(My.Resources.Resource.RunningError)
			context.Response.Write(eRunReport.ToString())
			Return
		End Try

		Dim primaryStream As MemoryStream = DirectCast(streams.GetPrimaryStream().OpenStream(), MemoryStream)
		context.Response.BinaryWrite(primaryStream.ToArray())
	End Sub

	Public ReadOnly Property IsReusable As Boolean Implements IHttpHandler.IsReusable
		Get
			Return True
		End Get
	End Property

	Private NotInheritable Class HtmlStreamProvider
		Inherits MemoryStreamProvider
		Protected Overrides Function GetStreamUri(extension As String) As Uri
			Dim uri As Uri = MyBase.GetStreamUri(extension)
			Return New Uri(uri.OriginalString + HandlerCacheExtension, UriKind.Relative)
		End Function
	End Class
End Class
